# นางสาว นิรชร บรรณสาร

# 603020474-9

# ข้อ1

## สร้าง function ชื่อ smartDict ซึ่งทำหน้าที่:
- function รับ input เป็น list 2 ตัวที่มีขนาดเท่ากัน โดย list ตัวแรกเป็น list ของตัวเลขlistที่สองเป็น list ของtext
- ถ้า ค่า list แรกเป็นเลขคี่  ให้หาค่าใน dict ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 1 และ value จาก list ตัวที่ 2
- ถ้า ค่า list แรกเป็นเลขคู่  ให้หาค่าใน dict ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 2 และ value จาก list ตัวที่ 1
- output ของ function คือ dict ที่สร้างขึ้น


เช่น 
- output1 = smartDict([1,2,3],['apple','banana','orange'])
- จะได้ output1 เป็น{1:'apple','banana':2,3:'orange'}

In [175]:
def   smartDict(listnumber,listtext):  
        if listnumber==  :        
                retdict ={ }
                for index in range(len(listnumber)) : 
                    retdict[listnumber[index]] = listtext[index]
        else:
                retdict ={ }
                for index in range(len(listnumber)) :
                     retdict[listtext[index]] = listnumber[index]
        return retdict
                 

In [ ]:
smartDict([1,2,3],['apple','banana','orange'])

# ข้อ 2

### จากไฟล์ Wuhan_nCov_GNome.text จงเขียนโปรแกรมเพื่อตอบคำถามต่อไปนี้

#### 2.1 กรดอะมิโนใด(A,B,G,C) มีปริมาณมากสุด และมีเท่าไหร่

In [9]:
import pandas as pd

In [10]:
DNAseq = pd.read_csv('C:\\Users\\Biw\\Documents\\viz_python\\python_dataviz\\Wuhan_nCoV_GNome.txt',header = None)

In [11]:
DNAseq.head()

0
0  ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGA...
1  CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTG...
2  TAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGC...
3  TTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCG...
4  CCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTT...

In [12]:
def countChar(seq,character): #row1,ตัอักษรแต่ละตัว
    charcount = 0 
    for char in seq:
        if char == character: 
            charcount = charcount + 1 # เอาตัวที่นับเจอไปบวกกับตัวนับไปแล้ว
    return charcount

In [15]:
ANum = 0 #หาความยาว
BNum = 0
GNum = 0
CNum = 0
for row in range (DNAseq.shape[0]):      #rang (len(DNAseq))
    ANum += countChar(DNAseq.iloc[row][0],'A')
    TNum += countChar(DNAseq.iloc[row][0],'T')
    GNum += countChar(DNAseq.iloc[row][0],'G')
    CNum += countChar(DNAseq.iloc[row][0],'C')

print(f'There are  {ANum} A, {TNum} B , {GNum} G , {CNum} C ')

There are  8954 A, 28782 B , 5863 G , 5492 C 


 ตอบ B มากสุด ทั้งหมด 28782 ตัว

### 2.2 sequence ที่มีขนาด 3 และ 4 ที่เกิดซ้ำมากที่สุด

### ซ้ำ3ตัว

In [121]:
def populatate3charstring(seq):
    str_list = list()
    for i in range(len(seq)-2) : 
        str_list.append(seq[i:i+3]) 
        
    return str_list

In [123]:
LIST_ALL_3CHAR = list() 
for row in range (DNAseq.shape[0]):
    LIST_ALL_3CHAR = LIST_ALL_3CHAR + populatate3charstring(DNAseq.iloc[row][0])

In [125]:
len(LIST_ALL_3CHAR)

29047

In [126]:
unique_3char = (set(LIST_ALL_3CHAR))

In [127]:
len(unique_3char)

64

In [130]:
dict_3char = {}
for sub_str in unique_3char: 
    num1 = 0
    for s in LIST_ALL_3CHAR: 
        if sub_str == s: 
            num1 += 1
    dict_3char[sub_str] = num1

In [131]:
dict_3char

{'TCT': 526,
 'CGA': 91,
 'CGT': 166,
 'GTT': 686,
 'GTC': 265,
 'GAC': 328,
 'CGC': 92,
 'GCC': 184,
 'GAT': 430,
 'ACT': 660,
 'ATT': 749,
 'TGT': 834,
 'CTG': 483,
 'TCA': 536,
 'CGG': 74,
 'CCG': 70,
 'TCG': 110,
 'TAG': 414,
 'ATC': 336,
 'GTA': 458,
 'CAA': 681,
 'ACG': 157,
 'TAC': 591,
 'CCT': 332,
 'GTG': 532,
 'AGA': 582,
 'CAT': 472,
 'TCC': 203,
 'TAT': 601,
 'CCC': 113,
 'CAG': 424,
 'CCA': 336,
 'AGC': 292,
 'GGC': 217,
 'CAC': 447,
 'ATA': 456,
 'AAC': 595,
 'ACA': 789,
 'ATG': 701,
 'TTA': 853,
 'GCT': 509,
 'TGG': 540,
 'TGA': 602,
 'TGC': 530,
 'GCG': 86,
 'GGA': 272,
 'TAA': 688,
 'CTT': 720,
 'CTC': 276,
 'AAT': 735,
 'AGG': 317,
 'AAG': 564,
 'ACC': 364,
 'GGT': 442,
 'TTG': 800,
 'GCA': 363,
 'CTA': 543,
 'TTT': 984,
 'GAG': 290,
 'GAA': 524,
 'AAA': 899,
 'TTC': 509,
 'GGG': 131,
 'AGT': 493}

In [132]:
maxVal1 = -100 
maxKey1 = None 
for i in dict_3char.keys(): 
    if dict_3char[i] > maxVal1: 
        maxVal1 = dict_3char[i]
        maxKey1 = i

In [134]:
maxKey1

'TTT'

In [135]:
print(f'max 3-long-sequence is {maxKey1} with frequency equals {maxVal1}')

max 3-long-sequence is TTT with frequency equals 984


#### ซ้ำ 4 ตัว

In [122]:
def populatate4charstring(seq):
    str2_list = list()
    for i in range(len(seq)-3) : 
        str2_list.append(seq[i:i+4]) 
        
    return str2_list

In [160]:
LIST_ALL_4CHAR = list() 
for row in range (DNAseq.shape[0]):
    LIST_ALL_4CHAR = LIST_ALL_4CHAR + populatate4charstring(DNAseq.iloc[row][0])

In [161]:
len(LIST_ALL_4CHAR)

28619

In [162]:
unique_4char = (set(LIST_ALL_4CHAR))

In [163]:
len(unique_4char)

256

In [164]:
dict_4char = {}
for sub2_str in unique_4char: 
    num2 = 0
    for s in LIST_ALL_4CHAR: 
        if sub2_str == s: 
            num2 += 1
    dict_4char[sub2_str] = num2

In [165]:
dict_4char

{'AGAT': 140,
 'CCAC': 78,
 'GATA': 77,
 'ATGG': 168,
 'TTGA': 185,
 'GGTC': 28,
 'TTAC': 235,
 'ACAT': 190,
 'GAAT': 117,
 'TTTC': 150,
 'CTTG': 160,
 'TACA': 231,
 'GCAC': 93,
 'AGCA': 83,
 'TAGA': 141,
 'GCCC': 16,
 'CAAC': 181,
 'TAAA': 247,
 'GCAG': 76,
 'CGAC': 14,
 'CATT': 152,
 'GGGA': 30,
 'CCTA': 99,
 'TGAA': 188,
 'TGTG': 180,
 'TCCG': 15,
 'GATG': 166,
 'GTAC': 128,
 'TACT': 199,
 'GACG': 28,
 'AGTA': 105,
 'ACGA': 33,
 'TGGC': 106,
 'TCCC': 26,
 'ACAC': 162,
 'ATAC': 104,
 'ACCA': 139,
 'GCTT': 158,
 'GGCG': 19,
 'CCGC': 19,
 'TCGG': 22,
 'GGGT': 55,
 'TTAG': 121,
 'GTCC': 43,
 'AATG': 220,
 'AACC': 106,
 'CTTC': 128,
 'CAGC': 69,
 'CCGG': 9,
 'AATA': 137,
 'CCGT': 26,
 'GAAA': 160,
 'ATCG': 21,
 'CACG': 39,
 'GAGT': 97,
 'GACC': 53,
 'ACGT': 61,
 'GCAA': 112,
 'AATC': 106,
 'ATGA': 179,
 'AAAG': 197,
 'AACG': 43,
 'AAAC': 182,
 'CTAG': 87,
 'GGTA': 118,
 'CTGG': 98,
 'CAAT': 169,
 'TTCT': 205,
 'CGCT': 27,
 'TGGA': 134,
 'GACA': 151,
 'CGCC': 13,
 'AGTT': 171,
 'GCGA': 19

In [166]:
maxVal2 = -100 
maxKey2 = None 
for i in dict_4char.keys(): 
    if dict_4char[i] > maxVal2: 
        maxVal2 = dict_4char[i]
        maxKey2 = i

In [167]:
maxKey2

'TGTT'

In [168]:
print(f'max 4-long-sequence is {maxKey2} with frequency equals {maxVal2}')

max 4-long-sequence is TGTT with frequency equals 318


# ข้อ3  จากไฟล์ data.xlxs สร้างตาราง

## 3.1

- ใหม่ที่มี2 columns โดย columns แรกชื่อว่า ภาษาอีสาน และมีค่าใน columns  เป็นค่าจาก columns  ภาษาอีสานเดิมแต่เปลี่ยนคำที่เป็น '/' เป็น ',' และ columns  ที่สองชื่อว่า ภาษาไทย และมีค่าใน columns เป็นค่าจาก columns  ภาษาไทยเดิมแต่เปลี่ยนค่าที่เป็น '/ ' 
- เช่นจากสวัสดี/สุผุสุคน/เด้อ/จ้า สวัสดี/ทุกๆ คน/นะ/คะ เป็น สวัสดี,สุผุสุคน,เด้อ,จ้า สวัสดี,ทุกๆ คน,นะ,คะ 

In [76]:
data= pd.read_excel('C:\\Users\\Biw\\Documents\\viz_python\\python_dataviz\\data.xlsx',header = None)

In [77]:
L = ['ภาษาอีสาน' , 'ภาษาไทย']
Data = pd.DataFrame.from_records([],columns = L)
for i in range(32): # 32 คือชีท
    df = pd.read_excel('data.xlsx', sheet_name = i, usecols = 'B:C', skipfooter = 1) #บีซี คือหัวคอลัมบีซีในexcel
    Data = pd.concat([Data,df], axis = 0, ignore_index = True) #ไม่สนใจอินเดก

In [78]:
Data.iloc[0]['ภาษาอีสาน'].replace("/", ",")

'สวัสดี,สุผู้สุคน,เด้อ,จ้า'

In [79]:
def replacestr(col1):
    co1 = col1.replace("/", ",")
    return co1

In [80]:
list_1 = []
list_2 = []
for i in range(len(Data)):
    list_1.append(replacestr(Data.iloc[i]['ภาษาอีสาน']))
    list_2.append(replacestr(Data.iloc[i]['ภาษาไทย']))

In [104]:
len(list_1)

975

In [105]:
len(list_2)

975

In [81]:
Data['ภาษาอีสานใหม่'] = list_1
Data['ภาษาไทยใหม่'] = list_2

In [82]:
Data

ภาษาอีสาน  \
0                            สวัสดี/สุผู้สุคน/เด้อ/จ้า   
1         กิน/ข้าว/ก่อน/เด้อ/จ้า/ ไผ/สิ/ถ่า/เค้า/ดาวน์   
2    เค้า/ดาว/กะ/ขอ/ให้/เค้า/ดาว/อย่าง/มี/ความ/สุข/...   
3    นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...   
4                       ขอ/ให้/เชิญ/แซ่บ/เด้อ/พี่/น้อง   
..                                                 ...   
970  นั่น/แหละ/คุย/กัน/ไป่/อูย/คุย/กัน/ไป่/เป็น/ปี/...   
971  กะ/เล๊ย/แบบ/ชี/วิต/ชวง/นั่น/แบบ/สิ/เอา/จัง/ได๋...   
972  เส็ด/แล่ว/กะ/เล๊ย/อ่า/เพิ่น/เซิน/ปะ/เออ/เซิน/ม...   
973  ซวง/นั่น/ข่อย/กะ/เขว/เนาะ/จ้า/สิ/ไป่/ตอ/หรือ/ส...   
974  ข่อย/กะ/เล๊ย/ตัด/สิน/ใจ/กั๊บ/มู/หว่า/เออ/ข่อย/...   

                                               ภาษาไทย  \
0                                สวัสดี/ทุก ๆ คน/นะ/คะ   
1            กิน/ข้าว/ก่อน/นะ/จ้ะ/ใคร/จะ/รอ/เค้า/ดาวน์   
2    เค้า/ดาวน์/ก็/ขอ/ให้/เค้า/ดาน์/อย่าง/มี/ความ/ส...   
3    นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...   
4                          ขอ/ให้/เชิญ/อร่อย/นะ/ทุก/คน   
..                                                 ...   
970  นั้น/แหละ/คุย/กัน/ไป/อูย/คุย/กัน/ไป/เป็น/ปี/อย...   
971  ก็/เลย/แบบ/ชี/วิต/ช่วง/นั้น/แบบ/จะ/เอา/ยัง/ไง/...   
972  เสร็จ/แล้ว/ก็/เลย/อ่า/เขา/เชิญ/ปะ/เออ/เชิญ/มา/...   
973  ช่วง/นั้น/ฉัน/ก็/เคว้ง/อ่ะ/ค่ะ/จะ/ไป/ต่อ/หรือ/...   
974  ฉัน/ก็/เลย/ตัด/สิน/ใจ/กับ/เพื่อน/ว่า/เออ/ฉัน/จ...   

                                         ภาษาอีสานใหม่  \
0                            สวัสดี,สุผู้สุคน,เด้อ,จ้า   
1         กิน,ข้าว,ก่อน,เด้อ,จ้า, ไผ,สิ,ถ่า,เค้า,ดาวน์   
2    เค้า,ดาว,กะ,ขอ,ให้,เค้า,ดาว,อย่าง,มี,ความ,สุข,...   
3    นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...   
4                       ขอ,ให้,เชิญ,แซ่บ,เด้อ,พี่,น้อง   
..                                                 ...   
970  นั่น,แหละ,คุย,กัน,ไป่,อูย,คุย,กัน,ไป่,เป็น,ปี,...   
971  กะ,เล๊ย,แบบ,ชี,วิต,ชวง,นั่น,แบบ,สิ,เอา,จัง,ได๋...   
972  เส็ด,แล่ว,กะ,เล๊ย,อ่า,เพิ่น,เซิน,ปะ,เออ,เซิน,ม...   
973  ซวง,นั่น,ข่อย,กะ,เขว,เนาะ,จ้า,สิ,ไป่,ตอ,หรือ,ส...   
974  ข่อย,กะ,เล๊ย,ตัด,สิน,ใจ,กั๊บ,มู,หว่า,เออ,ข่อย,...   

                                           ภาษาไทยใหม่  
0                                สวัสดี,ทุก ๆ คน,นะ,คะ  
1            กิน,ข้าว,ก่อน,นะ,จ้ะ,ใคร,จะ,รอ,เค้า,ดาวน์  
2    เค้า,ดาวน์,ก็,ขอ,ให้,เค้า,ดาน์,อย่าง,มี,ความ,ส...  
3    นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...  
4                          ขอ,ให้,เชิญ,อร่อย,นะ,ทุก,คน  
..                                                 ...  
970  นั้น,แหละ,คุย,กัน,ไป,อูย,คุย,กัน,ไป,เป็น,ปี,อย...  
971  ก็,เลย,แบบ,ชี,วิต,ช่วง,นั้น,แบบ,จะ,เอา,ยัง,ไง,...  
972  เสร็จ,แล้ว,ก็,เลย,อ่า,เขา,เชิญ,ปะ,เออ,เชิญ,มา,...  
973  ช่วง,นั้น,ฉัน,ก็,เคว้ง,อ่ะ,ค่ะ,จะ,ไป,ต่อ,หรือ,...  
974  ฉัน,ก็,เลย,ตัด,สิน,ใจ,กับ,เพื่อน,ว่า,เออ,ฉัน,จ...  

[975 rows x 4 columns]

In [83]:
datanew= Data[['ภาษาอีสานใหม่','ภาษาไทยใหม่']]
datanew.head()

ภาษาอีสานใหม่  \
0                          สวัสดี,สุผู้สุคน,เด้อ,จ้า   
1       กิน,ข้าว,ก่อน,เด้อ,จ้า, ไผ,สิ,ถ่า,เค้า,ดาวน์   
2  เค้า,ดาว,กะ,ขอ,ให้,เค้า,ดาว,อย่าง,มี,ความ,สุข,...   
3  นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...   
4                     ขอ,ให้,เชิญ,แซ่บ,เด้อ,พี่,น้อง   

                                         ภาษาไทยใหม่  
0                              สวัสดี,ทุก ๆ คน,นะ,คะ  
1          กิน,ข้าว,ก่อน,นะ,จ้ะ,ใคร,จะ,รอ,เค้า,ดาวน์  
2  เค้า,ดาวน์,ก็,ขอ,ให้,เค้า,ดาน์,อย่าง,มี,ความ,ส...  
3  นี่,จ้า,อา,หาร,เย็น,เป็น,ลา,ซาน,ย่า,กับ,ผัก,สด...  
4                        ขอ,ให้,เชิญ,อร่อย,นะ,ทุก,คน

## ข้อ3.2 

- สร้างตาราง Dictionary โดยตารางประกอบด้วยสอง columns คือ ภาษาอีสานและภาษาไทย โดยค่าใน columns ภาษาอีสาน คือคำในภาษาอีสาน และcolumns ภาษาไทย คือคำแปลเป็นภาษากลางของภาษาอีสานในcolumns แรก (ใช้ค่าจากตาราง data.text โดยคำที่อยู่ในลำดับเดียวกันของ columns ภาษาอีสานและภาษาไทย คือคำแปลของกันและกัน จำนวนคำของทั้งสอง columns เท่ากัน -ถ้าจำนวนคำระหว่างสอง columns ไม่เท่ากันให้ข้ามrecord นั้นไป)

In [84]:
def splitstr(col1):
    co1 = col1.split('/')
    return co1

In [85]:
splitstr(df.iloc[0]['ภาษาอีสาน'])

['ทัก',
 'ทาย',
 'กัน',
 'ซัมบายดี',
 'สวัสดี',
 'วัน',
 'อัง',
 'คาร',
 'เนาะ',
 'จ้า',
 'อา',
 'กาศ',
 'มื้อ',
 'นี่',
 'หนาว',
 'พี',
 'น้อง',
 'แล่ว',
 'ก่อ',
 'อ่า',
 'ช่วง',
 'หน่า',
 'หนาว',
 'นี่',
 'จ้า',
 'ทาง',
 'ทาง',
 'เมือง',
 'นอก',
 'กะ',
 'สิ',
 'แบบ',
 'หว่า']

In [91]:
list_11 = []
list_22 = []
for i in range(len(Data)):
    if len(splitstr(Data.iloc[i]['ภาษาอีสาน'])) == len(splitstr(Data.iloc[i]['ภาษาไทย'])): # ตัดก่อนแล้วค่อยดู
        list_11 += splitstr(Data.iloc[i]['ภาษาอีสาน']) # list_11 = list_11 + splitstr(Data.iloc[i]['ภาษาอีสาน'])
        list_22 += splitstr(Data.iloc[i]['ภาษาไทย'])

In [92]:
len(list_22)

30852

In [93]:
len(list_11)

30852

In [94]:
def google(col1,col2):
    str = f'{col1} แปลว่า {col2}'
    return str

In [169]:
A = []
for i in range(len(list_11)):
    if len(list_11[i]) == len(list_22[i]):
        A += [google(list_11[i],list_22[i])]

In [118]:
Aa = set(A)

In [120]:
Aa

{'สอ แปลว่า สอ',
 'ซึ่ง แปลว่า ซึ่ง',
 'องศา แปลว่า องศา',
 'เหลือง แปลว่า เหลือง',
 'หน่อย แปลว่า หน่อย',
 'เผ็ด แปลว่า หรือ',
 'เกิน แปลว่า เกิน',
 'เข็น แปลว่า เข็น',
 'ชิม แปลว่า ชิม',
 'ฮัก แปลว่า รัก',
 'เต้า แปลว่า เต้า',
 'หนะ แปลว่า อ่ะ',
 'วีดีโอ แปลว่า วีดีโอ',
 'เมนท์ แปลว่า เมนท์',
 'เย็บ แปลว่า เย็บ',
 'แป็บ แปลว่า แป็บ',
 'เว้ย แปลว่า เว้ย',
 'หนะ แปลว่า เลย',
 'เปลือก แปลว่า เปลือก',
 'จาน แปลว่า จาน',
 'ซอก แปลว่า ซอก',
 'เฉพาะ แปลว่า เฉพาะ',
 'ใฮ่ แปลว่า ให้',
 'กอ แปลว่า กอ',
 'พี้ แปลว่า นี่',
 'ศิลปะ แปลว่า ศิลปะ',
 'ฮอท แปลว่า ฮอท',
 'ทอง แปลว่า ทอง',
 'บ่ แปลว่า รึ',
 'เกรง แปลว่า เกรง',
 'คิด แปลว่า คิด',
 'คุ๊ณ แปลว่า คุ๊ณ',
 'ใหญ  แปลว่า ใหญ่',
 'เฮา แปลว่า ฉัน',
 'มก แปลว่า มก',
 'วื้ย แปลว่า ว้าย',
 'อนุญาต แปลว่า อนุญาต',
 'น้อย แปลว่า อยู่',
 'ถือ แปลว่า ถือ',
 'อ่าน แปลว่า อ่าน',
 'สา แปลว่า สา',
 'สอน แปลว่า สอน',
 'บ่ แปลว่า ม่',
 'ทาย แปลว่า ทาย',
 'ซาย แปลว่า ชาย',
 'รพ แปลว่า รพ',
 'ตื้น แปลว่า ตื้น',
 'จาก แปลว่า จาก',
 'รับ แปลว่า รับ',
 'เทอร์ แปล